src/Database/Query.php

```diff
@@ -408,80 +408,49 @@ public function traverseParts(callable $visitor, array $parts)
      *         ->from('articles');
      *
      *     return $cte
-     *         ->setName('cte')
-     *         ->setQuery($cteQuery);
+     *         ->name('cte')
+     *         ->query($cteQuery);
      * });
      * ```
      *
-     * The list of expressions can be reset by overwriting and passing `null` for the
-     * expression:
-     *
-     * ```
-     * $query->with(null, true);
-     * ```
-     *
-     * @param \Cake\Database\Expression\CommonTableExpression|\Closure|null $expression The CTE to add.
+     * @param \Closure|\Cake\Database\Expression\CommonTableExpression $cte The CTE to add.
+     * @param bool $recursive Whether the CTE is recursive.
      * @param bool $overwrite Whether to reset the list of CTEs.
      * @return $this
-     * @throws \InvalidArgumentException When passing `null` for the `$expression` argument but not enabling
-     *  `$overwrite`.
-     * @throws \InvalidArgumentException When an invalid type is passed or returned for the `$expression` argument.
-     * @throws \InvalidArgumentException When the given CTE object has no name set.
-     * @throws \InvalidArgumentException When the given CTE object has no query set.
-     * @throws \InvalidArgumentException When a CTE object with the same name is already attached to this query.
      */
-    public function with($expression, $overwrite = false)
+    public function with($cte, bool $recursive = false, bool $overwrite = false)
```

Comment:
Having recursive be a mode on the expression will be more readable. With this form a long with will have a trailing `true` and figuring out what that means will require IDE jump/hover. A named method doesn't have that drawback.